In [3]:
import pandas as pd
import numpy as np

import seaborn as sns
from matplotlib import pyplot as plt
%matplotlib inline

In [4]:
## Read the dataset
df = pd.read_csv('data/WA_Fn-UseC_-Telco-Customer-Churn.csv')
len(df)

FileNotFoundError: [Errno 2] No such file or directory: 'data/WA_Fn-UseC_-Telco-Customer-Churn.csv'

In [ ]:
df.head(5)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [ ]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

- **CustomerID**: the ID of the customer
- **Gender**: male/female
- **SeniorCitizen**: whether the customer is a senior citizen (0/1)
- **Partner**: whether they live with a partner (yes/no)
- **Dependents**: whether they have dependents (yes/no)
- **Tenure**: number of months since the start of the contract
- **PhoneService**: whether they have phone service (yes/no)
- **MultipleLines**: whether they have multiple phone lines (yes/no/no phone service)
- **InternetService**: the type of internet service (no/fiber/optic)
- **OnlineSecurity**: if online security is enabled (yes/no/no internet)
- **OnlineBackup**: if online backup service is enabled (yes/no/no internet)
- **DeviceProtection**: if the device protection service is enabled (yes/no/no internet)
- **TechSupport**: if the customer has tech support (yes/no/no internet)
- **StreamingTV**: if the TV streaming service is enabled (yes/no/no internet)
- **StreamingMovies**: if the movie streaming service is enabled (yes/no/no internet)
- **Contract**: the type of contract (monthly/yearly/two years)
- **PaperlessBilling**: if the billing is paperless (yes/no)
- **PaymentMethod**: payment method (electronic check, mailed check, bank transfer, credit card)
- **MonthlyCharges**: the amount charged monthly (numeric)
- **TotalCharges**: the total amount charged (numeric)
- **Churn**: if the client has canceled the contract (yes/no)


In [ ]:
df.dtypes


customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

In [ ]:
df.dtypes.value_counts()

object     18
int64       2
float64     1
Name: count, dtype: int64

In [ ]:
## Missing value in TotalCharges
total_charges =pd.to_numeric(df.TotalCharges,errors='coerce')
df[total_charges.isnull()][['customerID','TotalCharges']]

customerID TotalCharges
488   4472-LVYGI             
753   3115-CZMZD             
936   5709-LVOEQ             
1082  4367-NUYAO             
1340  1371-DWPAZ             
3331  7644-OMVMY             
3826  3213-VVOLG             
4380  2520-SGTTA             
5218  2923-ARZLG             
6670  4075-WKNIU             
6754  2775-SEFEE

In [ ]:
## Set the missing values to zero
df.TotalCharges = pd.to_numeric(df.TotalCharges, errors='coerce')
df.TotalCharges = df.TotalCharges.fillna(0)

In [ ]:
## columns uniform by lowercasing 
df.columns = df.columns.str.lower().str.replace(' ','_')
string_columns = list(df.dtypes[df.dtypes =='object'].index)

for col in string_columns:
    df[col] = df[col].str.lower().str.replace(' ','_')

In [ ]:
## target convert in boolean
df.churn = (df.churn == 'yes').astype(int)

In [ ]:
## data splitting
from sklearn.model_selection import train_test_split

In [ ]:
df_train_full, df_test = train_test_split(df,test_size=0.2, random_state=42)

In [ ]:
df_train, df_val = train_test_split(df_train_full, test_size=0.33, random_state=42)

In [ ]:
y_train = df_train.churn.values
y_val = df_val.churn.values

In [ ]:
del df_train['churn']
del df_val['churn']

### EDA Exploratory data aanalysis

In [ ]:
df_train_full.isnull().sum()

customerid          0
gender              0
seniorcitizen       0
partner             0
dependents          0
tenure              0
phoneservice        0
multiplelines       0
internetservice     0
onlinesecurity      0
onlinebackup        0
deviceprotection    0
techsupport         0
streamingtv         0
streamingmovies     0
contract            0
paperlessbilling    0
paymentmethod       0
monthlycharges      0
totalcharges        0
churn               0
dtype: int64

In [ ]:
## Distribution of values in the Target variable
df_train_full.churn.value_counts()

churn
0    4138
1    1496
Name: count, dtype: int64

In [ ]:
## The proportion of churned users
global_mean  = df_train_full.churn.mean()
round(global_mean,2)*100

27.0

the dataset is imbalanced. The churn rate is our date 27%

In [ ]:
categorical = ['gender', 'seniorcitizen', 'partner', 'dependents',
'phoneservice', 'multiplelines', 'internetservice',
'onlinesecurity', 'onlinebackup', 'deviceprotection',
'techsupport', 'streamingtv', 'streamingmovies',
'contract', 'paperlessbilling', 'paymentmethod']
numerical = ['tenure', 'monthlycharges', 'totalcharges']

In [ ]:
df_train_full[categorical].nunique()

gender              2
seniorcitizen       2
partner             2
dependents          2
phoneservice        2
multiplelines       3
internetservice     3
onlinesecurity      3
onlinebackup        3
deviceprotection    3
techsupport         3
streamingtv         3
streamingmovies     3
contract            3
paperlessbilling    2
paymentmethod       4
dtype: int64

### Feature importance analysis

What makes customers churn? and What are the characteristics of people who churn?

#### churn rate

In [ ]:
female_mean = df_train_full[df_train_full.gender == 'female'].churn.mean()
male_mean = df_train_full[df_train_full.gender == 'male'].churn.mean()
print(f'male is {round(male_mean*100,2)}% and female{round(female_mean*100,2)}% ')

male is 26.05% and 27.08% 


The difference between the group rates for both females
and males is quite small, which indicates that knowing the gender of the customer
doesn’t help us identify whether they will churn.

In [ ]:
partner_yes = df_train_full[df_train_full.partner == 'yes'].churn.mean()
partner_no = df_train_full[df_train_full.partner == 'no'].churn.mean()
print(f'"yes" is {round(partner_yes*100,2)}% and  "no" {round(partner_no*100,2)}% ')

"yes" is 20.07% and  "no" 32.64% 


As we see, the rates for those who have a partner are quite different from rates for
those who don’t: 20% and 33%, respectively. It means that clients with no partner are
more likely to churn than the ones with a partner

### Risk ratio

In [ ]:
df_group = df_train_full.groupby(by='gender').churn.agg(['mean'])
df_group

mean
gender          
female  0.270841
male    0.260478

In [ ]:
df_group['diff'] = df_group['mean'] - global_mean
df_group['risk'] = df_group['mean']/ global_mean
df_group

mean      diff      risk
gender                              
female  0.270841  0.005310  1.019998
male    0.260478 -0.005053  0.980971

In [ ]:
from IPython.display import display
for col in categorical:
    df_group = df_train_full.groupby(by=col).churn.agg(['mean'])
    df_group['diff'] = df_group['mean'] - global_mean
    df_group['risk'] = df_group['mean']/ global_mean
    display(df_group)
    

mean      diff      risk
gender                              
female  0.270841  0.005310  1.019998
male    0.260478 -0.005053  0.980971

mean      diff      risk
seniorcitizen                              
0              0.237098 -0.028433  0.892922
1              0.413907  0.148377  1.558793

mean      diff      risk
partner                              
no       0.326446  0.060916  1.229411
yes      0.200733 -0.064798  0.755968

mean      diff      risk
dependents                              
no          0.312326  0.046795  1.176233
yes         0.155674 -0.109856  0.586276

mean      diff      risk
phoneservice                              
no            0.253623 -0.011908  0.955156
yes           0.266824  0.001293  1.004871

mean      diff      risk
multiplelines                                 
no                0.251397 -0.014134  0.946771
no_phone_service  0.253623 -0.011908  0.955156
yes               0.284105  0.018574  1.069952

mean      diff      risk
internetservice                              
dsl              0.191851 -0.073679  0.722521
fiber_optic      0.415558  0.150028  1.565010
no               0.076606 -0.188924  0.288502

mean      diff      risk
onlinesecurity                                   
no                   0.416014  0.150484  1.566727
no_internet_service  0.076606 -0.188924  0.288502
yes                  0.145342 -0.120189  0.547363

mean      diff      risk
onlinebackup                                     
no                   0.398693  0.133162  1.501494
no_internet_service  0.076606 -0.188924  0.288502
yes                  0.216531 -0.048999  0.815467

mean      diff      risk
deviceprotection                                 
no                   0.387706  0.122175  1.460117
no_internet_service  0.076606 -0.188924  0.288502
yes                  0.226825 -0.038705  0.854234

mean      diff      risk
techsupport                                      
no                   0.413472  0.147941  1.557153
no_internet_service  0.076606 -0.188924  0.288502
yes                  0.152855 -0.112676  0.575657

mean      diff      risk
streamingtv                                      
no                   0.335418  0.069887  1.263197
no_internet_service  0.076606 -0.188924  0.288502
yes                  0.298945  0.033415  1.125841

mean      diff      risk
streamingmovies                                  
no                   0.333333  0.067803  1.255348
no_internet_service  0.076606 -0.188924  0.288502
yes                  0.301320  0.035789  1.134784

mean      diff      risk
contract                                    
month-to-month  0.426533  0.161002  1.606340
one_year        0.117987 -0.147544  0.444343
two_year        0.028379 -0.237151  0.106878

mean      diff      risk
paperlessbilling                             
no                0.16414 -0.101390  0.618159
yes               0.33594  0.070409  1.265164

mean      diff      risk
paymentmethod                                          
bank_transfer_(automatic)  0.174475 -0.091056  0.657080
credit_card_(automatic)    0.152404 -0.113126  0.573961
electronic_check           0.449921  0.184390  1.694420
mailed_check               0.190328 -0.075203  0.716782

- For **gender**, there is not much difference between females and males. Both means are approximately the same, and for both groups, the risks are close to 1.
- **Senior citizens** tend to churn more than nonseniors: the risk of churning is 1.53 for seniors and 0.89 for nonseniors.
- People with a **partner** churn less than people with no partner. The risks are 0.75 and 1.22, respectively.
- People who use **phone service** are not at risk of churning: the risk is close to 1, and there’s almost no difference with the global churn rate. People who don’t use phone service are even less likely to churn: the risk is below 1, and the difference with the global churn rate is negative. 
- Clients with no **tech support** tend to churn more than those who do.
- People with **monthly contracts** cancel the contract a lot more often than others, and people with **two-year contacts** churn very rarely.
